beetsplug/playlistquery.py

```diff
@@ -0,0 +1,70 @@
+# -*- coding: utf-8 -*-
+# This file is part of beets.
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+
+import os
+from beets.plugins import BeetsPlugin
+from beets.dbcore import FieldQuery, types
+from beets.util.confit import NotFoundError
+
+class PlaylistQuery(FieldQuery):
+    """Matches files listed by a playlist file.
+    """
+    relative_path = None
+    playlist_dir = None
+
+    def __init__(self, field, pattern, fast=True):
+        super(PlaylistQuery, self).__init__(field, pattern, fast)
+
+        playlist_file = pattern + '.m3u'
+        playlist_path = os.path.join(self.playlist_dir, playlist_file)
+
+        self.paths = []
+        with open(playlist_path, 'r') as f:
+            for line in f:
+                if line[0] == '#':
+                    # ignore comments, and extm3u extension
+                    continue
+                self.paths.append(os.path.normpath(
+                    os.path.join(self.relative_path, line.decode('utf-8').rstrip())
+                ))
+
+    def match(self, item):
+        return item.path.decode('utf-8') in self.paths
+
+
+class PlaylistType(types.String):
+    """Custom type for playlist query.
+    """
+    query = PlaylistQuery
+
+
+class PlaylistQueryPlugin(BeetsPlugin):
+    item_types = {'playlist': PlaylistType()}
+
+    def __init__(self):
+        super(PlaylistQueryPlugin, self).__init__()
+        self.register_listener('library_opened', self.library_opened)
+
+        PlaylistQuery.playlist_dir = self.config['playlist_dir'].as_filename()
```

Comment:
At the moment a Query object is constructed with the field and pattern, which makes it very clean. I think it is unwise to expand that in the default case, since most of the time it will go unused. However, I can see potential for the Builder pattern here - have a Plugin object instantiate Query objects that it controls.

So currently, `construct_query_part()` in dbcore.queryparse directly constructs the selected `query_class`. Instead, it should know which Plugin object corresponds to each query class, then calls on that object to construct the desired class. The default implementation in BeetsPlugin would do what `construct_query_part()` does at the moment (i.e. check if we have a FieldQuery, and construct in two different ways) but can be overridden in custom plugins to do more complex things, passing in other information if desired.